tls/s2n_client_key_exchange.c

```diff
@@ -49,136 +51,90 @@ static int s2n_rsa_client_key_recv(struct s2n_connection *conn)
-
-    /* Turn the pre-master secret into a master secret */
-    GUARD(s2n_prf_master_secret(conn, &pms));
-
-    /* Erase the pre-master secret */
-    GUARD(s2n_blob_zero(&pms));
-
-    /* Expand the keys */
-    GUARD(s2n_prf_key_expansion(conn));
-
-    /* Save the master secret in the cache */
-    if (s2n_allowed_to_cache_connection(conn)) {
-        GUARD(s2n_store_to_cache(conn));
-    }
-
+    conn->handshake.rsa_failed |= !s2n_constant_time_equals(client_protocol_version, shared_key->data, S2N_TLS_PROTOCOL_VERSION_LEN);
     return 0;
 }
 
-static int s2n_dhe_client_key_recv(struct s2n_connection *conn)
+int s2n_dhe_client_key_recv(struct s2n_connection *conn, struct s2n_blob *shared_key)
 {
     struct s2n_stuffer *in = &conn->handshake.io;
-    struct s2n_blob shared_key = {0};
 
     /* Get the shared key */
-    if (conn->secure.cipher_suite->key_exchange_alg->flags & S2N_KEY_EXCHANGE_ECC) {
-        GUARD(s2n_ecc_compute_shared_secret_as_server(&conn->secure.server_ecc_params, in, &shared_key));
-    } else {
-        GUARD(s2n_dh_compute_shared_secret_as_server(&conn->secure.server_dh_params, in, &shared_key));
-    }
-
-    /* Turn the pre-master secret into a master secret */
-    GUARD(s2n_prf_master_secret(conn, &shared_key));
-
-    /* Erase the pre-master secret */
-    GUARD(s2n_blob_zero(&shared_key));
-    GUARD(s2n_free(&shared_key));
-
-    /* Expand the keys */
-    GUARD(s2n_prf_key_expansion(conn));
+    GUARD(s2n_dh_compute_shared_secret_as_server(&conn->secure.server_dh_params, in, shared_key));
+    /* We don't need the server params any more */
+    GUARD(s2n_dh_params_free(&conn->secure.server_dh_params));
+    return 0;
+}
 
-    /* Save the master secret in the cache */
-    if (s2n_allowed_to_cache_connection(conn)) {
-        GUARD(s2n_store_to_cache(conn));
-    }
+int s2n_ecdhe_client_key_recv(struct s2n_connection *conn, struct s2n_blob *shared_key)
+{
+    struct s2n_stuffer *in = &conn->handshake.io;
 
+    /* Get the shared key */
+    GUARD(s2n_ecc_compute_shared_secret_as_server(&conn->secure.server_ecc_params, in, shared_key));
     /* We don't need the server params any more */
-    if (conn->secure.cipher_suite->key_exchange_alg->flags & S2N_KEY_EXCHANGE_ECC) {
-        GUARD(s2n_ecc_params_free(&conn->secure.server_ecc_params));
-    } else {
-        GUARD(s2n_dh_params_free(&conn->secure.server_dh_params));
-    }
-
+    GUARD(s2n_ecc_params_free(&conn->secure.server_ecc_params));
     return 0;
 }
 
 int s2n_client_key_recv(struct s2n_connection *conn)
 {
-    if (conn->secure.cipher_suite->key_exchange_alg->flags & S2N_KEY_EXCHANGE_DH) {
-        return s2n_dhe_client_key_recv(conn);
-    } else {
-        return s2n_rsa_client_key_recv(conn);
-    }
+    const struct s2n_key_exchange_algorithm *kem_core = conn->secure.cipher_suite->key_exchange_alg;
+    struct s2n_blob shared_key = {0};
+
+    GUARD(kem_core->client_key_recv(conn, &shared_key));
```

Comment:
Can you add an assert that kem_core->client_key_recv is not NULL? Need to check what the policy of the S2N library regarding asserts.